providers/utils.go

```diff
@@ -168,46 +168,60 @@ func (i *ProviderInfo) GetOption(key string) (string, bool) {
 	return val, ok
 }
 
-// ReadModuleInfo is a version of ReadModuleInfoWithErr without an error.
-// Deprecated.
-// TODO this is a temporary impl which relies on default values for erroneous cases.
+// ReadModuleInfo finds information about the module.
+// If module is not found fallbacks to the default.
 func (i *ProviderInfo) ReadModuleInfo(moduleID common.ModuleID) *ModuleInfo {
-	result, _ := i.ReadModuleInfoWithErr(moduleID)
-
-	return result
-}
+	// Empty value fallback to the default value defined in ProviderInfo.
+	if moduleID == "" {
+		moduleID = i.defaultModuleOrRoot()
+	}
 
-func (i *ProviderInfo) ReadModuleInfoWithErr(moduleID common.ModuleID) (*ModuleInfo, error) {
+	// RootModule is inferred using the core values of ProviderInfo.
+	// On the surface those connectors are module agnostic.
 	rootModule := ModuleInfo{
 		BaseURL:     i.BaseURL,
 		DisplayName: i.DisplayName,
 		Support:     i.Support,
 	}
 
-	// Root or empty module name should default to provider information.
-	if moduleID == common.ModuleRoot || len(moduleID) == 0 {
-		return &rootModule, nil
+	// No modules exist. Fallback to the one and only RootModule.
+	if i.Modules == nil || len(*i.Modules) == 0 {
+		if moduleID != common.ModuleRoot {
+			// TODO the catalog should be checked almost at the "compile time".
+			// TODO There should be tests to ensure integrity. When anything is changed it should do consistency check.
+			slog.Warn("provider doesn't have modules while a module was requested",
+				"provider", i.DisplayName, "modules", moduleID)
+
+			return &rootModule // common.ErrMissingModule
+		}
+
+		// Requesting root when no modules exist is allowed.
+		return &rootModule
 	}
 
-	if i.Modules == nil {
-		// Module ID was requested when no modules are defined on provider.
-		slog.Warn("modules info is not defined for provider while requested a module",
-			"provider", i.DisplayName, "modules", moduleID)
+	// Root module is providerInfo derived module.
+	if moduleID == common.ModuleRoot {
+		return &rootModule
+	}
 
-		return &rootModule, common.ErrMissingModule
+	// Find module.
+	module, ok := (*i.Modules)[moduleID]
+	if ok {
+		return &module
 	}
 
-	modules := *i.Modules
+	slog.Warn("module info is missing for a module",
+		"provider", i.DisplayName, "modules", moduleID)
 
-	module, ok := modules[moduleID]
-	if !ok {
-		slog.Warn("module info is missing for a module",
-			"provider", i.DisplayName, "modules", moduleID)
+	return &rootModule // common.ErrMissingModule
+}
 
-		return &rootModule, common.ErrMissingModule
+func (i *ProviderInfo) defaultModuleOrRoot() common.ModuleID {
```

Comment:
When no default module is set, this falls back to root, and we should log that if the provider has got modules - we would need to monitor on this